lib/chef/provider/package/apt.rb

```diff
@@ -127,11 +127,23 @@ def unlock_package(name, version)
 
         private
 
+        # compare 2 versions to each other to see which is newer.
+        # this differs from the standard package method because we
+        # need to be able to parse debian version strings which contain
+        # tildes which Gem cannot properly parse
+        #
+        # @return [Integer] 1 if v1 > v2. 0 if they're equal. -1 if v1 < v2
         def version_compare(v1, v2)
-          gem_v1 =  v1.gsub(/[_+]/, "+" => "-", "_" => "-") unless v1.nil?
-          gem_v2 =  v2.gsub(/[_+]/, "+" => "-", "_" => "-") unless v2.nil?
-
-          Gem::Version.new(gem_v1) <=> Gem::Version.new(gem_v2)
+          # we need to be defensive here since we can be passed a nil
+          dpkg_v1 = v1.nil? ? 0 : v1.to_s
+          dpkg_v2 = v2.nil? ? 0 : v2.to_s
+          if shell_out_compact_timeout("dpkg", "--compare-versions", dpkg_v1, "gt", dpkg_v2).status.success?
```

Comment:
this should raise if either v1 or v2 are nil, it makes no sense to have it silently translate it to a '0' here.

we already have a check that the candidate_version is not nil prior to calling this function in the superclass, and its likely that we need a check that the installed_version is not nil there.  and that is the appropriate place for it to go, which would short-circuit this test (if its not installed, and there's no installed_version it makes no sense to compare the versions so that use case should never fall through to this code).

that also makes the code simpler:  `shell_out_compact_timeout("dpkg", "--compare-versions", dpkg_v1.to_s, "gt", dpkg_v2.to_s).status.success?`

Reply:
when I tested it in dpkg nils were being passed in